### PR TITLE
Creation action class

### PR DESCRIPTION
**Pull request by Matt-Warnock**

### CODE REVIEW

File: lib/creator.rb [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+    @user_interface = user_interface
+  end
+
+  def run
+    loop do
+      contact_details = user_interface.ask_for_fields
+
+      database.create(contact_details)
+      user_interface.display(contact_details)
+      break if user_interface.add_another_contact? == false
```
Comment (octopusinvitro) [inline lib/creator.rb:15]: it's an antipattern to compare the result of a boolean method with a boolean, the boolean method is already returning a boolean.

If what you want is the opposite boolean, you can use `unless` instead of `if`:

```ruby
break unless user_interface.add_another_contact?
```

File: spec/creator_spec.rb [hunk 2]
```diff
@@ -0,0 +1,77 @@
+# frozen_string_literal: true
+
+require 'creator'
+
+RSpec.describe Creator do
+  describe '#run' do
+    let(:database) { double('ArrayDatabase', create: nil) }
```
Comment (octopusinvitro) [inline spec/creator_spec.rb:7]: In this case, the database is simple enough that I would use it directly:

```ruby
let(:database) { ArrayDatabase.new }
```

Try to use mocks only when needed. For example in the case of the user interface, you need it because you have to wait for user input and you don't have a human user entering things in automated tests.

File: spec/creator_spec.rb [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    let(:ui) do
+      double('UserInterface',
+             ask_for_fields: test_details,
+             display: nil,
+             add_another_contact?: false,
+             menu_choice: nil)
+    end
+
+    it 'tells the UI to ask for contact details' do
+      creator = described_class.new(ui, database)
```
Comment (octopusinvitro) [inline spec/creator_spec.rb:17]: You could extract this variable as most tests are using it :+1: 

```ruby
let(:creator) { described_class.new(ui, database) }
```

File: spec/creator_spec.rb [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+      allow(ui).to receive(:add_another_contact?).and_return(true, false)
+
+      creator = described_class.new(ui, database)
+
+      creator.run
+
+      expect(ui).to have_received(:ask_for_fields).twice
+      expect(database).to have_received(:create).twice
+      expect(ui).to have_received(:display).twice
+      expect(ui).to have_received(:add_another_contact?).twice
```
Comment (octopusinvitro) [inline spec/creator_spec.rb:58]: It's a good practice to have only 1 `expect` per test, except in very extremely edge cases.

In this case the most important thing that happens is that you add two contacts to the database, so I would write an expectation for that :+1:

**Review by octopusinvitro — CHANGES_REQUESTED** [review]
Good work :+1: 

Some changes to make though

File: spec/creator_spec.rb (hunk 2 above)
Comment (Matt-Warnock) [reply spec/creator_spec.rb:7]: The reason I made a double of the database is to test create being called.
I don't think I can do that without making a double of some kind.

File: spec/creator_spec.rb (hunk 4 above)
Comment (Matt-Warnock) [reply spec/creator_spec.rb:58]: How can I test that create has been called if it is not a double? 
The only way I can think of is to test an actual ArrayDatabase's array for the contact, but wouldn't I be testing the ArrayDatabase class if I did that?

Comment (Matt-Warnock) [reply spec/creator_spec.rb:58]: Would it be miss leading to test the ArrayDatabase class in the Creator class spec?

File: spec/creator_spec.rb (hunk 2 above)
Comment (octopusinvitro) [reply spec/creator_spec.rb:7]: you don't have to test that create was called
what does create do?
that is what you have to test :+1:

File: spec/creator_spec.rb (hunk 4 above)
Comment (octopusinvitro) [reply spec/creator_spec.rb:58]:
```ruby
expect(database.count).to eq(2)
```

Comment (Matt-Warnock) [reply spec/creator_spec.rb:58]: So, it is ok to test other class' as long as their in the chain of the described class?

File: spec/creator_spec.rb (hunk 2 above)
Comment (Matt-Warnock) [reply spec/creator_spec.rb:7]: "in this case....... that is what you have to test"
Does it take ages to learn all these rules? 
Seems like there are so many rules and with difficult correlations.

**Review by octopusinvitro — APPROVED** [review]
